library/Icingadb/Data/Nodes.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Icinga\Module\Icingadb\Data;
+
+use ArrayIterator;
+use Countable;
+use Icinga\Module\Icingadb\Common\Auth;
+use Icinga\Module\Icingadb\Common\Database;
+use Icinga\Module\Icingadb\Model\DependencyNode;
+use Icinga\Module\Icingadb\Model\Host;
+use ipl\Stdlib\Filter;
+use IteratorAggregate;
+
+class Nodes implements IteratorAggregate, Countable
+{
+    use Database;
+    use Auth;
+
+    protected $nodes;
+
+    /** @var Filter\Rule */
+    protected $filter;
+
+    public function __construct(Filter\Rule $filter)
+    {
+        $this->filter = $filter;
+    }
+
+    public function getIterator(): ArrayIterator
+    {
+        if ($this->nodes === null) {
+            $membersQuery = DependencyNode::on($this->getDb())
+                ->with([
+                    'host',
+                    'host.state',
+                    'service',
+                    'service.state',
+                    'service.host'
+                ])
+                ->filter($this->filter);
+
+            $this->applyRestrictions($membersQuery);
+
+            $nodes = [];
+            foreach ($membersQuery as $node) {
+                $nodes[] = $node->service_id ? $node->service : $node->host;
```

Comment:
`$node->service_id !== null` please, not sure if there's a chance that an id like 0x000… would ever occur, but this would be a false positive in such a case